### PR TITLE
[WIP] Remove the kindle5 target

### PR DESCRIPTION
**Pull request by onde2rock**
Remove the kindle target because it use the linaro TC instead of the custom build one.
It's now replaced with the kindle5 target which is equivalent.
For newer devices, kindlepw can be used and their might be some
perfomance increase.

See : https://github.com/koreader/koreader/issues/3878#issuecomment-381417577

TODO:

- [x] If everything ok, update the main repo (kodev and readme) see : https://github.com/koreader/koreader/pull/3902

- [x]  Apply`UBUNTU_COMPAT_CFLAGS` only on the linaro TC.

### CODE REVIEW

**Comment by Frenzie** [comment]
Pinging @NiLuJe for review

>  test to see if UBUNTU_COMPAT_CFLAGS is still needed.

I want to be able to compile a build for quick testing without requiring any special toolchains. I.e., the difference between quickly testing something and properly optimized. Especially on my laptop where I don't even have space for all those gigabytes of stuff. The official builds need to simultaneously target the widest possible range of devices coupled with the best optimizations. Requiring all that just to do some basic testing/development is not particularly helpful imo.

(Docker containers go a long way toward making that more workable but it's still harder and it takes up even more diskspace.)

**Comment by NiLuJe** [comment]
As for the potential downsides of always keeping the UBUNTU flags in, they range from minor to maybe-not-actually-a-downside. Let's break it down:

```-fno-finite-math-only```: Tone down ```-ffast-math``` a bit, which may not be a bad thing, since using ffast-math globally was a bit of a gamble to begin with ;). IIRC, very few things in our list of third-party libs ended up pulling the offending symbols anyway.

```-fno-stack-protector -U_FORTIFY_SOURCE```: Disables a bunch of stack smashing protection related stuff. Security vs. performance trade-off.
FWIW, everything except kindle-legacy defaults to this enabled.

**Comment by NiLuJe** [comment]
FWIW, when I mentioned the compiler's triplet:

1. Hope really hard that CHOST is set
2. Parse the output from ```gcc -dumpmachine```, which coonsists solely of the full triplet (i.e., ```arm-kobo-linux-gnueabi```)

**Comment by NiLuJe** [comment]
Missed a kindle TARGET in .travis.yml for Travis ;).

As for that SIGILL in the emulator build, err, lol? Is this a known issue?

**Comment by NiLuJe** [comment]
kodev will also need to be updated, but that's not in base ;).

**Comment by Frenzie** [comment]
Even CircleCI has the occasional mysterious segmentation fault. I can only assume it's caused by some kind of OOM interplay between the various virtualized containers. But Travis definitely craps out more.

To be fair, the stuff here in base is more demanding. More third-party dependencies that can fail to download, etc. Although I don't think I've ever seen CircleCI fail to compile base when it's updated.

File: Makefile.defs [hunk 1]
```diff
@@ -32,10 +32,9 @@ CMAKE=cmake
 
 # set this to your ARM cross compiler:
 # set CC CXX AR LD RANLIB
-ifneq (,$(filter $(TARGET), kindle arm-generic))
-    export KINDLE=1
+ifneq (,$(filter $(TARGET), arm-generic))
```
Comment (NiLuJe) [inline Makefile.defs]: Should probably drop the brain-twisty filter trick for getting a logical OR since there's now only one test ;).

```ifeq ($(TARGET), arm-generic)```

(And yeah, that syntax trick kinda makes my brain hurt, but I think you got it right in your latest commit ;)).

**Comment by Frenzie** [comment]
Case in point regarding what I said about not requiring special toolchains:

```
The following packages will be upgraded:
  gcc-4.8-base gcc-4.9-base libc6 libexpat1 libgcc1 libpython3.4-minimal
  libpython3.4-stdlib libstdc++6 perl perl-base perl-modules python3.4
  python3.4-minimal vim-common vim-tiny
15 upgraded, 226 newly installed, 0 to remove and 94 not upgraded.
Need to get 172 MB of archives.
After this operation, 540 MB of additional disk space will be used.
E: You don't have enough free space in /var/cache/apt/archives/.
```

That's inside a Docker container. (Even docker prune only gave me back 1.5 GB.)

**Comment by NiLuJe** [comment]
FWIW, with the four custom TCs installed:

```
du -h ~/x-tools
956M    /home/niluje/x-tools
```

**Comment by Frenzie** [comment]
I actually have a whopping 18 GB free on my laptop atm after an annoying bit of cleanup, but Docker really does a number on my other partition because *of course* it's not sticking anything in /home. :-P If I'm ever going to perform any upgrades on this laptop it'll most assuredly be the SSD. But even then more for performance reasons than for space, see [here](https://fransdejonge.com/2017/03/ssd-performance-in-modern-laptops/).

This 2016 laptop was more of a sidegrade, or whatever that's called. It's mostly a major upgrade compared to my 2010 netbook, although even that has much more disk space, but not exactly an upgrade to my '09 laptop. But it's so light and the battery lasts fairly long, at least with TLP and Powertop. And not while doing a lot of compiling or watching YouTube. (Except with youtube-dl. Play a video in mpv 5%, play it on YouTube closer to 50%. Either browsers are terrible at playing video or YouTube is. Probably both.) I wouldn't want to go back to a heavy one but I'd gladly give up some of that thinness to a proper network port.

tl;dr that was a long-winded way of saying that 1 GB had been reduced to irrelevancy by the late '00s but we've somehow managed to make a lack of disk space a thing again. Also I consciously chose this laptop model because they wayyyyy overcharge for a slightly larger SSD in those things.

**Comment by onde2rock** [comment]
It's weird, it's like travis didn't take into account the change I made to [.ci/build-script.sh](https://github.com/onde2rock/koreader-base/blob/c230aab3efedc4fe9673b7bf1d1c4f14c7933e67/.ci/build_script.sh#L21)
but is using everything else.

So it's builiding with `target=kindle` like before, but with the new makefile, where target kindle don't exist anymore.

**Comment by Frenzie** [comment]
We use TARGET=kindle everywhere.

https://github.com/koreader/koreader-base/blob/7a294b4b2625cea4bc2618f00b40256f5c14e4de/.travis.yml#L21

https://gitlab.com/koreader/nightly-builds/blob/d547b243ebff7a12283d1576b3e521f3d0de5427/.gitlab-ci.yml#L28

I'll worry about GitLab, but if you're changing the Makefile you also have to change Travis. ;-)

**Comment by NiLuJe** [comment]
If you have a bit of RAM, I highly recommend "cheating" by doing stuff in a tmpfs ;).

Knowing that Portage does its thing in /var/tmp and that /tmp should be what it says on the tin, I've taken advantage of that in my fstab:

```
 # TMPDIRs on tmpfs
tmpfs                   /tmp            tmpfs           defaults,noexec,nosuid,size=1G,mode=1777,noatime                0 0
tmpfs                   /var/tmp        tmpfs           defaults,size=10G,mode=1777,noatime                             0 0
```

Initially, the advantage of building in a tmpfs was mainly speed, but with the advent of SSD, you can also add "don't stress the SSD's wear leveling" ;).
It won't consume/reserve RAM unless needed, so it's fairly transparent ;).

**Comment by NiLuJe** [comment]
Good question, I only have the vaguest inkling of how Docker works, so, no idea ^^.

But yeah, I get it, on my end the laptop is the cheapest Macbook Air money could buy a few years ago, so, it's starting to show its age too... ^^.

File: .ci/build_script.sh [hunk 2]
```diff
@@ -13,12 +13,12 @@ if [ "$TARGET" = kobo ]; then
...
         -v "$(pwd):${DOCKER_HOME}/base" "${DOCKER_IMG}" \
         /bin/bash -c 'source /home/ko/.bashrc && cd /home/ko/base && sudo chown -R ko:ko . && make TARGET=kobo all'
-elif [ "$TARGET" = kindle ]; then
+elif [ "$TARGET" = kindle5 ]; then
     sudo chmod -R 777 "${HOME}/.ccache"
     docker run -t \
         -v "${HOME}/.ccache:${DOCKER_HOME}/.ccache" \
         -v "$(pwd):${DOCKER_HOME}/base" "${DOCKER_IMG}" \
-        /bin/bash -c 'source /home/ko/.bashrc && cd /home/ko/base && sudo chown -R ko:ko . && make TARGET=kindle5 all'
+        /bin/bash -c 'source /home/ko/.bashrc && cd /home/ko/base && sudo chown -R ko:ko . && make TARGET=kindle5 all; echo $PATH'
```
Comment (Frenzie) [inline .ci/build_script.sh]: Wouldn't it make more sense to add the debug at the beginning so you don't have to wait half an hour? :-P

Output:
```
/home/ko/x-tools/arm-kobo-linux-gnueabihf/bin:/home/ko/x-tools/arm-kindlepw2-linux-gnueabi/bin:/home/ko/x-tools/arm-kindle5-linux-gnueabi/bin:/home/ko/x-tools/arm-kindle-linux-gnueabi/bin:/home/ko/.luarocks/bin:/home/ko/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

Same Docker image output with TARGET=kindle5 from the nightly build here (with current base, obviously):
https://gitlab.com/koreader/nightly-builds/-/jobs/64268539

Don't really have time to look into this closer but maybe it'll help.

**Comment by onde2rock** [comment]
> Also, as written does this PR remove the ability to simply cross-compile with a regular (non-specialized) toolchain?
Yes

```
     HAS_KOBO_TC:=$(shell command -v arm-kobo-linux-gnueabihf-gcc 2> /dev/null)
     ifdef HAS_KOBO_TC
         CHOST?=arm-kobo-linux-gnueabihf
     else
         CHOST?=arm-linux-gnueabihf
```

Hum, I think having two TC for the same thing is asking for hard to debug bug. But If kobo is already doing this, why not ?

> Because even setting aside the issue of needing 15 GB disk space to create the Kindle TC issue (if it's an issue for me, the guy with 3+ TB of total space,[1] it's an issue for everyone), that greatly heightens the barrier to contributing.

I have 128G SSD, it's a pain, but doable.
Also the TC arm-linux-gnueabi doesn't exist on my distro (arch) , so I have to use docker anyway

My idea was that two TC were redundant, that's why I asked @NiLuJe the difference between the two, to see which one we should keep, but it could either, I don't have strong opinion on this.

**Comment by NiLuJe** [comment]
@Frenzie : Yeah, I just realized that too when you pushed the changes about the missing Kindle OTAs... >_<"

As for your concern about TC, no, it shouldn't, as long as we keep the UBUNTU-COMPAT bits in. You should be able to use *any* of the kindle* TARGETS with a Ubuntu and/or Linaro TC.

**Comment by onde2rock** [comment]
> As for your concern about TCs, no, it shouldn't: as long as we keep the UBUNTU_COMPAT bits in, you should be able to use any of the kindle* TARGETS with a Ubuntu and/or Linaro TC.
> 
> As I mentioned somewhere, you'd only need a triplet detection if you wanted to drop the UBUNTU compatriot flags when using the custom TC. And I'd recommend doing that via -dumpmachine vs. the ugly check done on the Kobo side ;).

Ok, I can try to do that.

**Comment by Frenzie** [comment]
@onde2rock 
> I have 128G SSD, it's a pain, but doable.

So do I, and it's not doable for me. It may or may not be if I didn't have Windows taking up roughly half of it. But it's not about if it's technically *doable* or not. It's about the same reason that I'm 10000% more likely to write a shellscript or a Lua/Python script than anything that requires compiling. If I were a Kindle user just randomly coming across this project I'd probably just give up if the change in this PR were merged.

On my desktop I have a 256 GB SSD dedicated to Linux. On that one it's more Docker's fault for choosing a ridiculous default location (/var/lib/docker) which I haven't partitioned to randomly be 20 GB larger than intended (even with over 6 GB of "just in case" wasted free space on the partition). Docker apparently doesn't do symlinks. But even so, on that 256 GB SSD I have a total of what, 20 GB free? Nothing there that's deletable. I mean, sure, I could delete the Android toolchain to make space for the Kindle toolchain but that would be patently ridiculous when I already have a generic ARM toolchain that works for everything, just isn't optimized. (It's already bad enough that I need a stupid special Android TC.)

The only reason it's doable in a realistic sense is because I have several 100 GB available and/or easily freeable on my 2 TB HDD. (Do I want to delete any of my GOG games? Heck no! I deleted the 12 GB I was using for my Steam Tomb Raider since I finished it and all.) But if this were just some random project I figured it might be cool to do some hacking on I would've never, *ever* have bothered with that kind of nonsense. :-)

@NiLuJe 
> As for your concern about TCs, no, it shouldn't: as long as we keep the UBUNTU_COMPAT bits in, you should be able to use any of the kindle* TARGETS with a Ubuntu and/or Linaro TC.

Excellent.

> TL;DR: So, yeah, the only obvious path left is simply to build the "kindle" target with the "kindle5" TC. (And hope we name things after the TARGET, and not the CHOST ;)).

The package name is something like koreader-target-chost (koreader-`kindlepw2` (target) -`arm-kindlepw2-linux-gnueabi` (chost))

**Comment by NiLuJe** [comment]
Then we may need to fudge things up a bit to ensure the OTA keep being named `kindle-` and not `kindle5-`?

File: Makefile.defs [hunk 3]
```diff
@@ -32,15 +32,26 @@ CMAKE=cmake
...
 # set CC CXX AR LD RANLIB
-ifneq (,$(filter $(TARGET), kindle arm-generic))
-    export KINDLE=1
+ifeq ($(TARGET), arm-generic)
     CHOST?=arm-linux-gnueabi
-else ifeq ($(TARGET), kindle5)
+else ifeq ($(TARGET), kindle)
     export KINDLE=1
-    CHOST?=arm-kindle5-linux-gnueabi
+    HAS_KINDLE5_TC:=$(shell command -v arm-kindle5-linux-gnueabi-gcc 2> /dev/null)
```
Comment (Frenzie) [inline Makefile.defs]: If I understand @NiLuJe 's suggestion correctly both this and the Kobo check would be better using something along these lines (untested):

```
ifeq ($(shell gcc -dumpmachine), arm-kindle5-linux-gnueabi)
```

(It's cleaner looking than the `command -v >/dev/null` surroundings in any case)

Comment (NiLuJe) [reply Makefile.defs]: Possibly even w/ $(CC) or whatever we're actually calling instead of hardcoding gcc ;).

**Comment by NiLuJe** [comment]
@Frenzie : Ah, that's good news, thanks for clearing that up ;).

File: Makefile.defs (hunk 3 above)
Comment (Frenzie) [reply Makefile.defs]: Good point! We also support clang, albeit mainly as an oddity for Mac OS. :-P

(It may or may not compile slightly faster, which would be conducive to testing, but the actual binaries are slightly slower. Also it might output clearer or at least different error messages when something fails to compile.)

PS Clang actually listens to gcc because it's a drop-in replacement. It pretends to be something like gcc 4.3.

File: Makefile.defs [hunk 4]
```diff
@@ -32,23 +32,31 @@ CMAKE=cmake
...
 # set CC CXX AR LD RANLIB
-ifneq (,$(filter $(TARGET), kindle arm-generic))
-    export KINDLE=1
+ifeq ($(TARGET), arm-generic)
     CHOST?=arm-linux-gnueabi
-else ifeq ($(TARGET), kindle5)
+else ifeq ($(TARGET), kindle)
     export KINDLE=1
-    CHOST?=arm-kindle5-linux-gnueabi
+    ifeq ($(shell arm-kindle5-linux-gnueabi-gcc -dumpmachine), arm-kindle5-linux-gnueabi)
```
[the remark this answers is not in the document]
Comment (NiLuJe) [reply Makefile.defs]: Yeah, no, I'm sorry, it's my bad. What I had in mind works if we were using $(CC) to devise $(CHOST), but we don't, we roughly do the opposite.

So I guess what was originally done (i.e;, checking if the custom TCs is in $PATH) is actually good enough.

My bad -_-".

[the remark this answers is not in the document]
Comment (onde2rock) [reply Makefile.defs]: Sure, no problem, I'm gonna revert the last commit.

[the remark this answers is not in the document]
Comment (NiLuJe) [reply Makefile.defs]: Note that this technically works, too, but then to get it to fail someone would have to be twisted enough to rename the TC's triplet ;p.

File: Makefile.defs [hunk 5]
```diff
@@ -32,15 +32,26 @@ CMAKE=cmake
...
+ifeq ($(TARGET), arm-generic)
     CHOST?=arm-linux-gnueabi
-else ifeq ($(TARGET), kindle5)
+else ifeq ($(TARGET), kindle)
     export KINDLE=1
-    CHOST?=arm-kindle5-linux-gnueabi
+    HAS_KINDLE5_TC:=$(shell command -v arm-kindle5-linux-gnueabi-gcc 2> /dev/null)
+    ifdef HAS_KINDLE5_TC
+        #Use the custom build TC
+        CHOST?=arm-kindle5-linux-gnueabi
```
Comment (NiLuJe) [inline Makefile.defs]: Random note (this is fine): Using `:=` instead of `?:` would mean enforce using the custom TC in case something else is set in the env's CHOST.

I actually prefer always honoring the env, and this is what we previously did, so this is perfectly fine ;).

File: Makefile.defs [hunk 6]
```diff
@@ -228,23 +239,26 @@ else ifeq ($(TARGET), ubuntu-touch)
...
+	MACHINE_KINDLE:=$(shell PATH='$(PATH)' $(CC) -dumpmachine 2>/dev/null)
+    ifeq ($MACHINE_KINDLE, arm-linux-gnueabi)
+        COMPAT_CFLAGS:=$(UBUNTU_COMPAT_CFLAGS)
+        COMPAT_CXXFLAGS:=$(UBUNTU_COMPAT_CFLAGS)
+    endif
 else ifeq ($(TARGET), kindlepw2)
 	ARM_ARCH:=$(ARMV7_A9_ARCH)
 	ARM_ARCH+=-mfloat-abi=softfp
-else ifeq ($(TARGET), kindle-legacy)
+    ifeq ($MACHINE_KINDLE, arm-linux-gnueabi)
```
[the remark this answers is not in the document]
Comment (NiLuJe) [reply Makefile.defs]: Ah, I see where you got the PATH thing from ;).

Another solution is to move that MACHINE assignment up, before the TARGET dance (somewhere after we set $(CC)), and reuse it in your tests ;).

[the remark this answers is not in the document]
Comment (Frenzie) [reply Makefile.defs]: Oh whoops, you're right. It's mysterious-ish but not *that* mysterious. :-P

[the remark this answers is not in the document]
Comment (NiLuJe) [reply Makefile.defs]: And in that case, I'd *definitely* want to assign the MACHINE var with `:=`, we do *NOT* want a mismatch between the actual -dumpmachine of the compiler we'll use and a stray MACHINE var in our env.

[the remark this answers is not in the document]
Comment (NiLuJe) [reply Makefile.defs]: I'm done for that round of comments ;).

[the remark this answers is not in the document]
Comment (NiLuJe) [reply Makefile.defs]: (I have no strong preference on which way to go, but since we already re-use MACHINE in multiple places, it probably makes a tiny bit more sense to do the same).

**Comment by NiLuJe** [comment]
LGTM! :+1:

**Comment by onde2rock** [comment]
Alright, I did one last read and a quick build test, should be good to go.
